### PR TITLE
Runner docker volumes

### LINKED ISSUE #799: runner: Docker socket mounts
Would be nice to auto-mount the Docker socket (in the GitLab runner in particular).
Might circumvent some Docker-in-Docker issues.

Alternatively:

- [x] expose `--docker-volumes` #861
- [ ] document `--docker-volumes` https://github.com/iterative/cml.dev/pull/189
- [x] add `docker_volumes` to TPI https://github.com/iterative/terraform-provider-iterative/pull/367

### PR DESCRIPTION
**Pull request by DavidGOrtega**
- closes #799 
 - ~~adds docker inside images to support dind~~
 - fixes a small warning log of unknown runtime  
 - adds GL macos runner

Usage

```js
cml runner --repo XXX --token XXX  --docker-volumes /test:/test --docker-volumes /test:/testb
```

### CODE REVIEW

File: src/drivers/gitlab.js
Comment (casperdcl) [inline src/drivers/gitlab.js:203]: was this a bog btw?

Comment (DavidGOrtega) [reply src/drivers/gitlab.js:203]: a warning in the logs

File: Dockerfile
Comment (casperdcl) [inline Dockerfile:60]: Not sure we should do this by default. Maybe leave it to users to do? How much does it increase image size by?

Comment (DavidGOrtega) [reply Dockerfile:60]: not much indeed, we have already discussed to include it.

Comment (DavidGOrtega) [reply Dockerfile:60]: I will check and tell you

File: get-docker.sh
Comment (0x2b3bfa0) [inline get-docker.sh:1]: Did you mean to `.gitignore` this file? 🤔

File: src/drivers/gitlab.js
Comment (0x2b3bfa0) [inline src/drivers/gitlab.js:12]: Wouldn't it be better to add it as a [hidden] command-line option instead of accessing `process.env` from many places?

https://github.com/iterative/cml/blob/53e1547abfc349b8e75ee71f02709b331db2e59c/bin/cml/runner.js#L413

Comment (DavidGOrtega) [reply src/drivers/gitlab.js:12]: Yes, i said this in slack, this way is to have a fast access to test. I had this code long time ago. But definitely we must expose the param (can be visible) and prepare the equivalent pr in tpi

Comment (casperdcl) [reply src/drivers/gitlab.js:12]: Yes and don't think we need GL/gitlab in the CLI option's name. We can raise a "warning: ignored" if it's specified with an unsupported provider.

Comment (DavidGOrtega) [reply src/drivers/gitlab.js:12]: We are on the same page

File: Dockerfile
Comment (DavidGOrtega) [reply Dockerfile]: From 1.76Gb to 2.14Gb @casperdcl

**Comment by 0x2b3bfa0** [comment]
Nitpick: should it be `--docker-volume` instead of `--docker-volumes`? 🤔

File: bin/cml/runner.js
Comment (0x2b3bfa0) [inline bin/cml/runner.js:422]:
```suggestion
        description: 'Docker volumes. This feature is only supported in GitLab'
```

File: get-docker.sh
Comment (0x2b3bfa0) [reply get-docker.sh:1]: Still seeing this file. 🤔

Comment (DavidGOrtega) [reply get-docker.sh]: I have removed it but testing the image I to discover the size I added it again my mistake

image

**Review by 0x2b3bfa0 — APPROVED** [review]
Looks good to me after a cursory static read. 🎉

Let's wait for @casperdcl to resolve the DinD question before merging, though.

**Comment by DavidGOrtega** [comment]
> Looks good to me after a cursory static read. 🎉
> 
> Let's wait for @casperdcl to resolve the DinD question before merging, though.

I have tested it successfully, please have a look to `voltest.txt` which is a test volume added in the runner. The file contains 

```
Hi there
```

```sh
Using docker image sha256:965e8d79d5f222af32d4bf5989c00b3f491cc6754af39cbd57f9552e763adcca for iterativeai/cml:0-dvc2-base1 with digest iterativeai/cml@sha256:a0ec4ad7663b9cc4699d8a085339c7810f7161130dfe9a2cfadcc0715f103969 ...
$ cat /voltest.txt
Hi there
$ curl -fsSL https://get.docker.com -o get-docker.sh && sh get-docker.sh
```

File: Dockerfile
Comment (0x2b3bfa0) [reply Dockerfile]: Given that we support DinD (Docker in Docker), it would be great to provide it as part of our container images.

Still ~380 MB are a bit too much. 🧐 Are we including the daemon also? If we aren't running it, we should probably install only the command-line tool.

Comment (0x2b3bfa0) [reply Dockerfile]: This doesn't serve the same purpose as [`docker:dind`](https://hub.docker.com/layers/docker/library/docker/dind/images/sha256-5ed82558a7494a410eb6777aef81cf618118517d352b34307ca40d1da8589a22) and just bloats our images.

If we want to include [`dind`] functionality in our container images, we need to follow the same steps as [`dind/Dockerfile`](https://github.com/docker-library/docker/blob/8baa881aab85f8398d2edbbcc0da4bd1f556dd98/20.10/dind/Dockerfile) and run `dockerd` through an special wrapper at the beginning of the entrypoint.

If we only want to ship the `docker` (client) command–line tool, it should be enough to download [`docker-20.10.9.tgz`](https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz) from the releases server and install the client binary, 52 MB.

Comment (0x2b3bfa0) [reply Dockerfile]: @iterative/cml, should we discuss this separately to avoid blocking this pull request?

Comment (casperdcl) [reply Dockerfile]: yes 380MB seems much...

```suggestion
RUN apt-get install -yqq apt-transport-https ca-certificates curl \
 && curl -fsSL "https://download.docker.com/linux/ubuntu/gpg" | gpg --dearmor --yes -o /usr/share/keyrings/docker-archive-keyring.gpg \
 && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu xenial stable" > /etc/apt/sources.list.d/docker.list \
 && apt-get update -qq -o Dir::Etc::sourcelist="sources.list.d/docker.list" -o Dir::Etc::sourceparts="-" -o APT::Get::List-Cleanup="0" \
 && apt-get install -yqq --no-install-recommends docker-ce-cli docker-scan-plugin docker-ce
```

(don't think we need `docker-ce-rootless-extras`)

Comment (DavidGOrtega) [reply Dockerfile]: Using the script is what docker recommends to install docker even within docker

Comment (0x2b3bfa0) [reply Dockerfile]: > _Using the script is what docker recommends to install docker even within docker_
[wikipedian_protester](https://xkcd.com/285)

File: bin/cml/runner.js
Comment (casperdcl) [inline bin/cml/runner.js:422]: can we raise an error for GH & BB users?

File: Dockerfile
Comment (0x2b3bfa0) [inline Dockerfile:62]: ### Follow-up of https://github.com/iterative/cml/pull/861#discussion_r787766420

Why don't we leave our containers without Docker for now? If we wish, we can create a follow–up issue to discuss what we need and how to integrate it.

```suggestion
```

File: Dockerfile
Comment (0x2b3bfa0) [reply Dockerfile]: Resolved in favor of https://github.com/iterative/cml/pull/861/files#discussion_r789262613

File: Dockerfile
Comment (0x2b3bfa0) [reply Dockerfile:62]: @iterative/cml

**Review by 0x2b3bfa0 — APPROVED** [review]

**Review by casperdcl — APPROVED** [review]

**Pull request merged**
